proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+
+Given that the ``splice`` and ``quote`` annotations are useful for human readers
+understanding how code is staged, it seems worthwhile to make them explicit.
+
+Of course, nothing prevents development of a tool that helps users insert
+``splice`` and ``quote`` annotations into their modules as part of a migration
+to using ``ExplicitLevelImports``.
+
+
+Syntactic alternatives
+######################
+
+There are several proposals for the syntax of explicit level imports:
+
+* The splice/quote modifier could be placed after the module name, e.g. ``import
+  M splice``, like qualified imports under ``ImportQualifiedPost`` (see
+  `proposal #190 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0190-module-qualified-syntax.rst>`_).
+  This could be the only option, or it could be an optional alternative to
+  ``import splice M``. Putting the keywords after the module name would make it
+  easier to align and sort import lists.
+
+* Using a pragma rather than a syntactic modifier would fit in better with
+  how ``SOURCE`` imports work and make writing backwards compatible code easier::
+
+    import {-# SPLICE #-} B
+
+* Some have objected that the ``import splice`` suggestion is ungrammatical,
+  unlike ``import qualified`` or ``import hiding``.
+
+  One possible alternative is ``$(import Foo)`` to represent a splice import, but this
+  syntax clashes with the existing syntax for declaration splices and significantly
+  changes the structure of the import syntax.
+
+  Another alternative suggested was ``import for splice``, which restores the
+  grammatical nature of the import.
+
+* The keywords ``splice`` and ``quote`` are different lengths, which interferes
+  with alignment.  Alternatively ``quote`` could be replaced with ``quoted``,
+  which is the same length as ``splice``.
+
+* The syntax does not provide a way to explicitly import at level 0; this is
+  indicated by the absence of a keyword. We could add a keyword for this, e.g.
+  ``default`` or ``target`` (although neither of these are ideal). It would also
+  be possible for a single import to refer to multiple levels simultaneously,
+  e.g. ``import M default, splice`` or
+  ``import Prelude qualified splice as SP (id, ($)), quote as QP (const), default (..)``.
+
+* Modifier syntax (see `proposal #370
+  <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0370-modifiers.rst>`_)
+  could be used, although it would seem inconsistent with the existing syntax
+  that mainly uses keywords (except for `{-# SOURCE #-}` imports).
+
+
+Implicit Splice/Quote Prelude imports
+#####################################
+
+In the proposal ``Prelude`` must be imported explicitly at non-zero levels.
+
+Another possible design would be to automatically import ``Prelude`` at all
+levels rather than just level 0.
+
+For us, it is undesirable to automatically add these additional imports and
+hence dependencies on certain stages unless they were actually used.
```

Comment:
I think it is mostly a matter of taste, since if the dependencies are added implicitly, one can always use `NoImplicitPrelude` to avoid them anyway. But a thing to think about is whether adding an explicit `import Prelude` at one level should *suppress* implicit imports at other levels, as discussed in [implicit `Prelude` imports](https://github.com/ghc-proposals/ghc-proposals/blob/4177423c8e7d5642fa2898cfc154af3792dbe3ba/proposals/0000-splice-imports.rst#implicit-prelude-imports).

If `Prelude` is implicitly in scope at all levels, and adding any explicit import suppresses all of them, then adding `import Prelude` may result in splices/quotes having scope errors due to `Prelude` no longer being in scope at other levels. So probably explicit imports should suppress only the implicit `Prelude` import at that level? But then it may also be surprising that `import Prelude hiding (foo)` still leaves `foo` available in quotes/splices...

So the current design is deliberately more explicit. It's certainly arguable that having to write `import quote Prelude`/`import splice Prelude` is more noisy than necessary, though, and we could go the other way here.

Reply:
This is good for build performance. Imagine if I am working on GHC and some library that uses TH. I would like smaller recompiled by default if I modify `Prelude`.  This applies with multi-package repl or the `ghc -M` that Hadrian uses, in principle.

Also good for cross, I don't want to have to build native things if I don't need them. Right now you don't need a stage 2 *native* base when using a stage 1 cross compiler. Let's keep it that way unless one opts in.

Reply:
> Since `Prelude` will necessarily be available at all stages

Bottom line is with ABI versions, cross compilation, and cross-package incremental compilation, that is *not* true.

Reply:
I updated the proposal with an argument along the lines @adamgundry suggests.